[Testing:Developer] Docs for Cypress e2e testing (#322)
Submitty/submitty.github.io: _docs/developer/testing/cypress.md

---
title: End-to-End testing with Cypress
---

[Cypress](https://www.cypress.io/) is an open source end-to-end testing framework. It runs tests against 
a locally hosted instance of Submitty by automating a browser. Tests are written in JavaScript and are executed 
in the Cypress test-runner.

## Installation 

Make sure you have `node.js` version 10 or higher installed locally. You can check using the command: `node -v` 

```bash
# from the site directory
npm install --include=dev
```

> If you are on Linux you may need the following dependencies if they are not already installed:
> ```bash
> apt-get install libgtk2.0-0 libgtk-3-0 libgbm-dev libnotify-dev libgconf-2-4 libnss3 libxss1 libasound2 libxtst6 xauth xvfb
> ```

Reviewer: Let's link to the official docs here so that we're not maintaining this list.

```suggestion
If you are on Linux, view [this page](https://docs.cypress.io/guides/getting-started/installing-cypress.html#Linux) to see what dependencies you may need to install additionally for Cypress.
```

GET /home 200 287.217 ms - -
GET /authentication/login 200 2.297 ms - -
POST /authentication/check_login?old=http%253A%252F%252Flocalhost%253A1501%252Fhome 302 2772.070 ms - 96
GET /home 302 210.507 ms - 210
GET /authentication/login?old=http%3A%2F%2Flocalhost%3A1501%2Fhome 200 264.544 ms - -
GET /authentication/login 200 1.629 ms - -
POST /authentication/check_login?old=http%253A%252F%252Flocalhost%253A1501%252Fhome 302 205.710 ms - 96
```

A complete overview can be found [here](https://docs.cypress.io/guides/core-concepts/test-runner.html#Overview)

### Running Headless

You can run the entire test suite in headless mode. This will pipe the browser display directly to Cypress without getting displayed and is faster than the test-runner. This can be done with the command:

```bash
# from the site directory
npx cypress run 
```

If any test fails a screen shot will be taken of the step that failed. Additionally, a video will be recorded each time of the run. They can be found under the `site/cypress/screenshots` and `site/cypress/videos` directories.

See the full usage [here](https://docs.cypress.io/guides/tooling/plugins-guide.html#Use-Cases)

> ## Writing Tests

Reviewer: Worth linking to https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests.html#Folder-Structure?

```
cypress.json
cypress/
├── fixtures
│   └── example.json
├── integration
│   └── ...
│
├── plugins
│   └── index.js
├── screenshots
│ 
├── support
│   ├── commands.js
│   └── index.js
└── videos
   
```

While the main test files are in the integration directory, there are some supporting files. The `fixtures` directory contains files used to mock data. This can be text files, JSON, images, etc. 

The `commands.js` file 
allow you to program Cypress commands. Actions that require multiple steps and might be done often can be grouped together under a single command to prevent repetition, for example the action of logging into the Submitty site is a command. You can also write plain JavaScript functions in this file and import them using ES6 module syntax.

The `support/index.js` file instructs Cypress which files to include with every test, you most likely won't need to touch this file. 

The `plugins/index.js` allows you to extend Cypress's default behavior. You can read more about them [here](https://docs.cypress.io/guides/tooling/plugins-guide.html#Use-Cases). 

The top level `cypress.json` is the global configuration used.

### Test Structure

> Cypress borrow's Mocha's Behavior-driven development (BDD) [test interface](https://mochajs.org/#bdd) to organize tests. 

Reviewer: ```suggestion
Cypress has adopted [Mocha's behavior-driven development (BDD) syntax](https://docs.cypress.io/guides/references/bundled-tools.html#Mocha) to organize tests.
```

> You will typically use `describes`, `context`, and `it` to organize tests. `describes` and `context` organize tests into groups, either or can be used they both act the same. Here is an example:

Reviewer: ```suggestion
You will typically use `describes` and `it` to organize tests. `describes` organize tests into groups, either or can be used they both act the same.
To test assertions, you will use the integrated [Chai tool](https://docs.cypress.io/guides/references/bundled-tools.html#Chai) to write your assertions. Here, you will use the `expect()` function wrapped around an object, and use a series of [assertions](https://docs.cypress.io/guides/references/assertions.html#BDD-Assertions) about it.

Here is an example:
```

Author: You mean `describes` and `context` right, `it` definitely has a different behavior with cypress, you can't write any commands or assertions inside of a `describes` or `context` block, only within an `it`

```js
describe('foo', () => {
	context('bar', () => {
		it('baz', () => {
			expect(true).to.equal(true);
		});
	});
});
```

This will produce the following structure in the test runner:

![Cypress organization in the test runner](/images/cypress-structure.png)

You can also use the `afterEach` and `beforeEach` function calls to perform setup and teardown for each test. 
Be sure to read Cypress's best-practices for using these functions [here](https://docs.cypress.io/guides/references/best-practices.html#Using-after-or-afterEach-hooks).

### Workflow 

A test will typically:

1. Navigate to a page
2. Find an element
3. *(Optionally) Interact with it*

> 4. *(Optionally) Make an assertion*

Reviewer: We should remove the `(Optionally)` bit here. Test cases that just go to a page and/or find an element should be the exception to the majority of pages that do interactions and make assertions.

4. [Assertions](https://docs.cypress.io/guides/references/assertions.html#Chai)

Cypress allows you to select elements through [JQuery Selectors](https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Cypress-is-Like-jQuery). 

You can also use the "selector playground" in the test runner which allows you to highlight part of the screen
and let Cypress generate a selector for you. *Note:* for deeply nested HTML elements you will most likely have to create the selector yourself.

![Cypress selector-playground example](/images/selector-playground.png)

You can then use the `cy.get()` function to attempt to find an element on the page. If your test only requires an element to exist on the page, using a `cy.get()` is a strong enough check. 

If an element is not immediately ready Cypress will automatically retry for 4000ms by default, this can be useful for elements that dynamically render or preventing flaky-ness.

**NOTE**

Most Cypress functions are asynchronous and yield a result instead of returning it, this requires you to chain
together actions.

```js
cy.get('my_element')
.then((result) => {
	//do something with the element we found

> });

Reviewer: Can we express this example using the async / await syntax? We should be encouraging it to avoid having huge promise chains if we can help it.

Author: So cypress specifically doesn't allow async/await, since it clashes with their API design / doesn't work.

From https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Asynchronous

>  Why can’t I use async / await?
>
> If you’re a modern JS programmer you might hear “asynchronous” and think: why can’t I just use async/await instead of learning some proprietary API?
>
> Cypress’s APIs are built very differently from what you’re likely used to: but these design patterns are incredibly intentional. We’ll go into more detail later in this guide.

This is probably due to Cypress implementing a version of something that behaves similar to promises but not exactly 
https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Not-Promises

Reviewer: Can we add a note and link to their docs on this?
